experimental/sev_guest/src/crypto.rs

```diff
@@ -0,0 +1,155 @@
+//
+// Copyright 2022 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+//! This module provides an optional implementation for encrypting and decrypting guest requests
+//! using the RustCrypto `aes-gcm` crate.
+
+use core::mem::size_of;
+
+use aes_gcm::{
+    aes::{
+        cipher::{BlockSizeUser, Unsigned},
+        Aes256,
+    },
+    AeadInPlace, Aes256Gcm, KeyInit, Nonce, Tag,
+};
+use zerocopy::{AsBytes, FromBytes};
+
+use crate::guest::{
+    AeadAlgorithm, GuestMessage, GuestMessageHeader, Message, CURRENT_HEADER_VERSION,
+    CURRENT_MESSAGE_VERSION,
+};
+
+/// Wrapper for encrypting and decrypting guest messages.
+///
+/// The sequence number is used as an initialization vector/nonce for AES-GCM. The sequence number
+/// is internally incremented on every encryption to avoid reuse. It is also incremented on every
+/// decryption to synchronize it with the sequence number used by the Platform Secure Processor
+/// (PSP).
+///
+/// To make sure that the same sequence number is never reused with the same key, the same key must
+/// never be used to instantiate more than one instance of this struct.
+///
+/// If a request fails the PSP will not increment the sequence number. This means that we will be
+/// out of sync with its sequence number. There is no safe way to recover from that, so the best
+/// course of action is to delete the key or shut down the VM.
+///
+/// See Chapter 7 and Section 8.26.2 in <https://www.amd.com/system/files/TechDocs/56860.pdf>.
+pub struct GuestMessageEncryptor {
+    cipher: Aes256Gcm,
+    sequence_number: u64,
+}
+
+impl GuestMessageEncryptor {
+    /// Creates a new instance.
+    ///
+    /// For now we always assume that VMPCK_0 will be used as the key.
+    pub fn new(key: &[u8]) -> Result<Self, &'static str> {
```

Comment:
Where is this key taken from?

Reply:
This is normally provided by the platform secure processor (PSP) in the secrets page that is injected into the guest memory at startup when running with AMD SEV-SNP.